responsemanager/allocator/allocator.go

```diff
@@ -0,0 +1,273 @@
+package allocator
+
+import (
+	"context"
+	"errors"
+
+	pq "github.com/ipfs/go-ipfs-pq"
+	peer "github.com/libp2p/go-libp2p-peer"
+)
+
+type Allocator struct {
+	ctx             context.Context
+	totalMemoryMax  uint64
+	perPeerMax      uint64
+	total           uint64
+	nextAllocIndex  uint64
+	messages        chan allocationRequest
+	peerStatuses    map[peer.ID]*peerStatus
+	peerStatusQueue pq.PQ
+}
+
+func NewAllocator(ctx context.Context, totalMemoryMax uint64, perPeerMax uint64) *Allocator {
+	return &Allocator{
+		ctx:             ctx,
+		totalMemoryMax:  totalMemoryMax,
+		perPeerMax:      perPeerMax,
+		total:           0,
+		peerStatuses:    make(map[peer.ID]*peerStatus),
+		peerStatusQueue: pq.New(makePeerStatusCompare(perPeerMax)),
+		messages:        make(chan allocationRequest, 16),
+	}
+}
+
+func (a *Allocator) AllocateBlockMemory(p peer.ID, amount uint64) <-chan error {
+	responseChan := make(chan error, 1)
+	done := make(chan struct{}, 1)
+	select {
+	case <-a.ctx.Done():
+		responseChan <- errors.New("context closed")
+	case a.messages <- allocationRequest{p, amount, allocOperation, responseChan, done}:
+	}
+	select {
+	case <-a.ctx.Done():
+	case <-done:
+	}
+	return responseChan
+}
+
+func (a *Allocator) ReleaseBlockMemory(p peer.ID, amount uint64) error {
+	responseChan := make(chan error, 1)
+	select {
+	case <-a.ctx.Done():
+		responseChan <- errors.New("context closed")
+	case a.messages <- allocationRequest{p, amount, deallocOperation, responseChan, nil}:
+	}
+	select {
+	case <-a.ctx.Done():
+		return errors.New("context closed")
+	case err := <-responseChan:
+		return err
+	}
+}
+
+func (a *Allocator) ReleasePeerMemory(p peer.ID) error {
+	responseChan := make(chan error, 1)
+	select {
+	case <-a.ctx.Done():
+		responseChan <- errors.New("context closed")
+	case a.messages <- allocationRequest{p, 0, deallocPeerOperation, responseChan, nil}:
+	}
+	select {
+	case <-a.ctx.Done():
+		return errors.New("context closed")
+	case err := <-responseChan:
+		return err
+	}
+}
+
+func (a *Allocator) Start() {
+	go func() {
+		a.run()
+		a.cleanup()
```

Comment:
nit: Any reason not to move this into `run()` as a defer? (then just invoke `go a.run()` here)